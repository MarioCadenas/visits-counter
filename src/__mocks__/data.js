export const mockData = {
  loadingTrue: {
    loading: true,
  },
  loadingFalse: {
    loading: false,
  },
  emptyVisits: {
    visits: []
  },
  notEmptyVisits: {
    visits: [{ id: '123', count: 5 }]
  }
};
export const mockFunctions = {
  createVisit: jest.fn(() => ({ data: { createVisit: { count: 1 } } })),
  updateVisit: jest.fn(() => ({ data: { updateVisit: { count: mockData.notEmptyVisits.visits[0].count + 1 } } })),
};
