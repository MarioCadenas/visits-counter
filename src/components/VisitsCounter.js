import React, { Component } from 'react';

export default class VisitsCounter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      id: props.visit ? props.visit.id : null,
      count: props.visit ? props.visit.count : 0
    }
  }

  async componentDidMount() {
    const { props: { updateVisit, createVisit }, state: { id, count: currentCount } } = this;

    if (!id) {
      const { data: { createVisit: { count } } } = await createVisit({ variables: { count: currentCount + 1 } });
      return this.setState({ count });
    }

    const { data: { updateVisit: { count } } } = await updateVisit({ variables: { id } });
    return this.setState({ count });
  }

  render() {
    const { state: { count } } = this;
    return <h1>{`Visitas totales: ${count}`}</h1>
  }
}
