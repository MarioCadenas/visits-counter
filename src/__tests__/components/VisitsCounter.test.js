import React from 'react';
import { shallow, mount } from 'enzyme';
import VisitsCounter from '../../components/VisitsCounter';
import { mockData, mockFunctions } from '../../__mocks__/data';

describe('Visits Counter component', () => {
   it('should correctly render Visits Counter component', () => {
    const { notEmptyVisits: { visits: [visit] } } = mockData;
    const { createVisit, updateVisit } = mockFunctions;
    const wrapper = shallow(
      <VisitsCounter
        visit={visit}
        createVisit={createVisit}
        updateVisit={updateVisit}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h1').getElement()).toEqual(<h1>Visitas totales: 5</h1>);
  });
  it('should call createVisit when empty visit', () => {
    const { updateVisit, createVisit } = mockFunctions;
    mount(
      <VisitsCounter
        visit={undefined}
        createVisit={createVisit}
        updateVisit={updateVisit}
      />
    );
    expect(createVisit).toHaveBeenCalledWith({ variables: { count: 1 } });
    expect(createVisit).toHaveReturnedWith({ data: { createVisit: { count: 1 } } });
  });
  it('should call updateVisit when not empty visit', () => {
    const { notEmptyVisits: { visits: [visit] } } = mockData;
    const { updateVisit, createVisit } = mockFunctions;
    mount(
      <VisitsCounter
        visit={visit}
        createVisit={createVisit}
        updateVisit={updateVisit}
      />
    );
    expect(updateVisit).toHaveBeenCalledWith({ variables: { id: visit.id } });
    expect(updateVisit).toHaveReturnedWith({ data: { updateVisit: { count: visit.count + 1 } } });
  });
});
