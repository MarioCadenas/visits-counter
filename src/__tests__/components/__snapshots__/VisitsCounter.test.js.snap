// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visits Counter component should correctly render Visits Counter component 1`] = `
<h1>
  Visitas totales: 5
</h1>
`;
